src/__tests__/posthog-core.js

```diff
@@ -677,24 +678,128 @@ describe('_loaded()', () => {
+            delete window.performance
+
+            try {
+                given('config', () => ({
+                    capture_pageview: true,
+                    capture_performance: true,
+                }))
+
+                given.subject()
+
+                expect(given.overrides.capture).toHaveBeenCalledWith(
+                    '$pageview',
+                    { performance: { navigation: [], paint: [], resource: [] } },
+                    { send_instantly: true }
+                )
+                expect(performance.getEntriesByType).not.toHaveBeenCalled()
+            } finally {
+                window.performance = performance
+            }
+        })
+
+        it('safely attempts to capture pageview with performance when enabled but getEntriesByType is not available in browser', () => {
+            const performance = { ...window.performance }
+            delete window.performance.getEntriesByType
+
+            try {
+                given('config', () => ({
+                    capture_pageview: true,
+                    capture_performance: true,
+                }))
+
+                given.subject()
+
+                expect(given.overrides.capture).toHaveBeenCalledWith(
+                    '$pageview',
+                    { performance: { navigation: [], paint: [], resource: [] } },
+                    { send_instantly: true }
+                )
+                expect(performance.getEntriesByType).not.toHaveBeenCalled()
+            } finally {
+                window.performance = performance
+            }
+        })
+
+        it('safely attempts to capture performance if a type of entry is not available in a browser', () => {
+            // e.g. IE does not implement performance paint timing
+            // https://developer.mozilla.org/en-US/docs/Web/API/PerformancePaintTiming
+            // even though it implements getEntriesByType
+
+            const performance = { ...window.performance }
+            delete window.performance.getEntriesByType
+
+            window.performance.getEntriesByType = jest.fn().mockImplementation((type) => {
+                if (type === 'paint') {
+                    throw new Error('IE does not implement this')
+                } else {
+                    return []
+                }
+            })
+
+            try {
+                given('config', () => ({
+                    capture_pageview: true,
+                    capture_performance: true,
+                }))
+
+                given.subject()
+
+                expect(given.overrides.capture).toHaveBeenCalledWith(
+                    '$pageview',
+                    { performance: { navigation: [], paint: [], resource: [] } },
+                    { send_instantly: true }
+                )
+                expect(performance.getEntriesByType).not.toHaveBeenCalled()
+            } finally {
+                window.performance = performance
```

Comment:
Your tests have really messy setup code. We can clean this up really nicely by leveraging beforeEach/afterEach is for. Tests should only test what's relevant.

Suggestion:

```
beforeEach(() => {
  jest.spyOn(window.performance, 'getEntriesByType').mockImplementation(() => given.performanceEntries())
})
```

And use given in each of your tests.

In general having try/catch in your tests is a definite smell.

Reply:
have added an extra level of describe to move some of the setup and teardown out of the tests

we can't spy on performance `getEntriesByType` directly (because of a JSDOM issue https://github.com/jsdom/jsdom/issues/3309) so have followed the advice here: https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom

Reply:
You can simplify this further by doing this in your definition then:

```js
Object.defineProperty(window, 'performance', {
    writable: true,
    value: {
        getEntriesByType: jest.fn().mockImplementation(() => given.performanceEntries()),
    }
})
```

This would clean up the `'safely attempts to capture performance if a type of entry is not available in a browser'` test.